src/webgl/p5.RendererGL.Retained.js

```diff
@@ -320,18 +193,144 @@ p5.RendererGL.prototype.drawBuffersScaled = function(
+    var src = model[bd.src];
+    if (src) {
+      // check if we need to create the GL buffer
+      var createBuffer = !buffer;
+      if (createBuffer) {
+        // create and remember the buffer
+        geometry[bd.dst] = buffer = gl.createBuffer();
+      }
+      // bind the buffer
+      gl.bindBuffer(gl.ARRAY_BUFFER, buffer);
+
+      // check if we need to fill the buffer with data
+      if (createBuffer || model.dirtyFlags[bd.src] !== false) {
+        var map = bd.map;
+        // get the values from the model, possibly transformed
+        var values = map ? map(src) : src;
+
+        // fill the buffer with the values
+        this._bindBuffer(buffer, gl.ARRAY_BUFFER, values);
+
+        // mark the model's source array as clean
+        model.dirtyFlags[bd.src] = false;
+      }
+      // enable the attribute
+      shader.enableAttrib(attr, bd.size);
+    } else {
+      if (buffer) {
+        // remove the unused buffer
+        gl.deleteBuffer(buffer);
+        geometry[bd.dst] = null;
+      }
+      // disable the vertex
+      gl.disableVertexAttribArray(attr.index);
+    }
+  }
+};
+
+p5.RendererGL.prototype._renderStroke = function(geometry, stroke) {
+  // check if the stroke is enabled
+  if (!this._doStroke || geometry.lineVertexCount < 2 || this._enableNormal) {
+    return;
+  }
+
+  var gl = this.GL;
+
+  stroke.bindShader();
+
+  // set the uniform values
+  stroke.setUniform('uStrokeColor', this._strokeColor);
+  stroke.setUniform('uStrokeWeight', this._strokeWeight);
+
+  // prepare the render buffers
+  this._prepareBuffers(geometry, stroke, strokeBuffers);
+
+  // render the stroke
+  this._applyColorBlend(this._strokeColor);
+  gl.drawArrays(gl.TRIANGLES, 0, geometry.lineVertexCount);
+
+  stroke.unbindShader();
 };
 
-p5.RendererGL.prototype._drawElements = function(drawMode, gId) {
-  this.GL.drawElements(
-    drawMode,
-    this.gHash[gId].numberOfItems,
-    this.GL.UNSIGNED_SHORT,
-    0
+p5.RendererGL.prototype._renderFill = function(geometry, fill, drawMode) {
+  // check if the fill is enabled
+  if (!this._doFill || geometry.vertexCount < 3) {
+    return;
+  }
+
+  var gl = this.GL;
+
+  fill.bindShader();
+
+  // TODO: optimize
+  fill.setUniform('uAmbientColor', this._ambientColor);
+  fill.setUniform('uMaterialColor', this._diffuseColor);
+  fill.setUniform('uSpecularColor', this._specularColor);
+  fill.setUniform('isTexture', !!this._tex);
+  if (this._tex) {
+    fill.setUniform('uSampler', this._tex);
+  }
+
+  var pointLightCount = this.pointLightColors.length / 3;
+  fill.setUniform('uPointLightCount', pointLightCount);
+  fill.setUniform('uPointLightLocation', this.pointLightPositions);
+  fill.setUniform('uPointLightColor', this.pointLightColors);
+
+  var directionalLightCount = this.directionalLightColors.length / 3;
+  fill.setUniform('uDirectionalLightCount', directionalLightCount);
+  fill.setUniform(
+    'uDirectionalLightDirection',
+    this.directionalLightDirections
```

Comment:
this feels a little redundant and is quite a long name. can we avoid changing this and just note with an inline comment in the shader?

Reply:
I think uLightDirection is more concise and direct (hehe pun intended). :) I personally value that over consistency in this particular instance, but this isn't one of the more pressing issues in this patch.